text/0000-road-to-typescript.md

```diff
@@ -0,0 +1,160 @@
+
+<!--
+To add when opening the RFC:
+
+**Note:** *this* RFC is not an appropriate place for discussing any of these specific concerns. Rather, discussion should focus on:
+
+- whether Ember should or should not aim to support TypeScript as a first-class citizen
+- if it should, whether there are any other parts of the roadmap which have been missed
+
+-->
+
+### Semantic Versioning and Supported TypeScript Versions
+
+Ember and TypeScript have fundamentally different views on Semantic Versioning (SemVer).
+
+Ember has a deep commitment to minimizing breaking changes in general, and to strictly following SemVer when breaking changes *are* made. (The use of lockstep versioning for Ember Data and Ember CLI complicates this commitment to a substantial degree, but that complication is outside the scope of this RFC.)
+
+TypeScript explicitly *does not* follow SemVer. TypeScript's core team argues that *every* change to the compiler is a breaking change, and that SemVer is therefore meaningless. (We do not agree with this characterization, but are also uninterested in arguing it. This RFC takes the TypeScript team's position as a given.) Accordingly, every TypeScript point release may be a breaking change, and "major" numbers for releases signify nothing beyond having reached `x.9` in the previous cycle.
+
+For TypeScript to be a first-class citizen of the Ember ecosystem, we need:
+
+- a policy defining what constitutes a breaking change for consumers of a library which publishes types, including Ember’s core libraries
+- tooling to detect breaking changes in types—whether from refactors, or from new TypeScript releases—and to minimize the amount of churn from breaking changes in TypeScript
+- a general and widely-adopted policy for supported TypeScript versions
+
+Once all three of those elements are adopted, end users will be able to have equally high confidence in the stability of published types as they do in their runtime code.
+
+## How we teach this
+
+This RFC intentionally defers most of the analysis of teaching work to an RFC dedicated to the question. At a high level, once the roadmap is complete, the Ember blog should announce support, and Ember’s docs should explicitly state that TypeScript is officially supported and show TypeScript examples alongside JavaScript examples.
+
+It is critical that these materials (including the blog post) emphasize that TypeScript support is *additive*, not a replacement for JavaScript, as discussed above in [Non-Goals](#non-goals).
+
+## Drawbacks
+
+Adding TypeScript support imposes an additional maintenance burden on all contributors to Ember:
+
+- documentation must be kept up to date in both JavaScript and TypeScript
+- changes to APIs must be kept in sync with published types
```

Comment:
The above two points only apply to cases where the underlying Ember source has not been updated to TypeScript and we're maintaining separate `d.ts` files, correct?

Reply:
Certainly that is true for the second case. (I currently anticipate that will be a matter of years for a lot of the core types, though.) For documentation, it is a mix: you are correct as far as API docs go, but for guides, tutorials, etc., the concern will remain regardless.

Reply:
@wagenet I tweaked the wording here to clarify the difference between implementation-derived types and hand-authored types in 19ffd9f – thanks for calling that out!